contents/blog/hog-story.md

```diff
@@ -0,0 +1,206 @@
+Queries are necessary but not sufficient for such a platform.
+
+You also need to be able to customize logic and interactions with other systems. This would address a problem nearly every PostHog customer encounters transforming and exporting data.
+
+Deeper in *Crafting Interpreters,* Marius knew, lay the key.
+
+The naive solution to giving developers customizable behavior would be letting them write JavaScript. But there are so many issues with running arbitrary JavaScript at scale.
+
+" We had multiple problems with our JavaScript-based dynamic code execution engine. Security was the major one that just killed it at the end. But the other big problem was that JavaScript is kind of unpredictable. You start a script, and it can just leak promises everywhere," Marius told me.
+
+This just wouldn’t work for code that could be triggered by any individual user event. “ There's some callback that's going to resolve like two hours later and you have no idea that it's out there, eating memory. You do that for tens of thousands of events every minute… That's going to crash.”
+
+This wasn't speculative, either. Our own internal, JavaScript-based plugin engine had these problems all the time. It was impossible to expose the same power to external developers and also keep the product online.
+
+JavaScript was limiting the scope of our ambition.
+
+# What's a bytecode, anyway?
+
+The first phase of our custom language, HogQL, was comparatively simple. The queries are interpreted according rules, then converted into *different* queries. It "runs" by poking the database and returning data.
+
+But Hog, a language that would support logic, loops, making requests, and restructuring data, would need an independent environment from the database to do its work.
+
+So Hog is compiled into *bytecode*, a series of instructions that run in a *virtual machine*, our HogVM. This gives us an environment for execution whose behavior we could completely specify and control.
+
+The HogVM is implemented in both Python and TypeScript. Someday, if we need even more performance, it may be built again in Rust.
+
+Because we control the language and the environment where it runs, we can do some interesting voodoo:
+
+The state of any HogVM instance can be paused, written to a file, and resumed later. This gives us complete control of code execution. Not just the definition of valid or invalid behaviors, but also where and when code runs.
+
+The immediate fruits of this control? [Data pipelines](), where developers can write completely custom logic for integrating PostHog with other tools.
+
+Despite this small but free-form development sandbox, we don’t have to worry that some errant code will pile up and bring down the services that run it. We’ve designed the entire stack from top to bottom to address the precise problem in front of us.
+
+It can’t do everything, but it can do what our customers need to manipulate and export data really well.
+
+# The power to ignore resistance
```

Comment:
I also think the section is too long, may "Now, thanks to HogQL and Hog, PostHog gained:" could be its own or folded in the final section